Failed to parse '.arsc' file: IOException: Config size < 28

Log level: Error
Version: 1.4.4
Binary: [jadx-gui-1.4.4-no-jre-win.exe](https://github.com/skylot/jadx/releases/download/v1.4.4/jadx-gui-1.4.4-no-jre-win.exe)
File causing the error: [resources.arsc.zip](https://github.com/skylot/jadx/files/9479118/resources.arsc.zip)
Java VM:
```
openjdk version "18.0.1.1" 2022-04-22
OpenJDK Runtime Environment (build 18.0.1.1+2-6)
OpenJDK 64-Bit Server VM (build 18.0.1.1+2-6, mixed mode, sharing)
```


jadx Error:

```
ERROR: Failed to parse '.arsc' file
jadx.core.utils.exceptions.JadxException: Error decode: D:\download\browser\resources.arsc
	at jadx.api.ResourcesLoader.decodeStream(ResourcesLoader.java:78)
	at jadx.core.dex.nodes.RootNode.loadResources(RootNode.java:173)
	at jadx.api.JadxDecompiler.load(JadxDecompiler.java:122)
	at jadx.gui.JadxWrapper.open(JadxWrapper.java:65)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.IOException: Config size < 28
	at jadx.core.xmlgen.ResTableParser.parseConfig(ResTableParser.java:373)
	at jadx.core.xmlgen.ResTableParser.parseTypeChunk(ResTableParser.java:221)
	at jadx.core.xmlgen.ResTableParser.parsePackage(ResTableParser.java:162)
	at jadx.core.xmlgen.ResTableParser.decodeTableChunk(ResTableParser.java:110)
	at jadx.core.xmlgen.ResTableParser.decode(ResTableParser.java:86)
	at jadx.core.xmlgen.ResDecoder.decode(ResDecoder.java:28)
	at jadx.core.dex.nodes.RootNode.lambda$loadResources$7(RootNode.java:173)
	at jadx.api.ResourcesLoader.decodeStream(ResourcesLoader.java:61)
	... 6 common frames omitted
```

_Originally posted by @oxou in https://github.com/skylot/jadx/issues/178#issuecomment-1235626628_

--- a/jadx-core/src/main/java/jadx/core/xmlgen/ParserConstants.java
+++ b/jadx-core/src/main/java/jadx/core/xmlgen/ParserConstants.java
@@ -28,9 +28,12 @@ public class ParserConstants {
 	protected static final int RES_XML_LAST_CHUNK_TYPE = 0x017f;
 	protected static final int RES_XML_RESOURCE_MAP_TYPE = 0x0180;
 
-	protected static final int RES_TABLE_PACKAGE_TYPE = 0x0200;
-	protected static final int RES_TABLE_TYPE_TYPE = 0x0201;
-	protected static final int RES_TABLE_TYPE_SPEC_TYPE = 0x0202;
+	protected static final int RES_TABLE_PACKAGE_TYPE = 0x0200; // 512
+	protected static final int RES_TABLE_TYPE_TYPE = 0x0201; // 513
+	protected static final int RES_TABLE_TYPE_SPEC_TYPE = 0x0202; // 514
+	protected static final int RES_TABLE_TYPE_LIBRARY = 0x0203; // 515
+	protected static final int RES_TABLE_TYPE_OVERLAY = 0x0204; // 516
+	protected static final int RES_TABLE_TYPE_STAGED_ALIAS = 0x0206; // 517
 
 	/**
 	 * Type constants

--- a/jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java
+++ b/jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java
@@ -150,7 +150,7 @@ public class ResTableParser extends CommonBinaryParser implements IResParser {
 		PackageChunk pkg = new PackageChunk(id, name, typeStrings, keyStrings);
 		resStorage.setAppPackage(name);
 
-		chunkLoop: while (is.getPos() < endPos) {
+		while (is.getPos() < endPos) {
 			long chunkStart = is.getPos();
 			int type = is.readInt16();
 			LOG.trace("res package chunk start at {} type {}", chunkStart, type);
@@ -174,9 +174,7 @@ public class ResTableParser extends CommonBinaryParser implements IResParser {
 					throw new IOException(
 							String.format("Encountered unsupported chunk type TYPE_STAGED_ALIAS at offset 0x%x ", chunkStart));
 				default:
-					// In some resources.arsc files endPos is not correctly set (e.g. points to the end of the file),
-					// so we have to stop when we encounter an unknown type entry
-					break chunkLoop;
+					LOG.warn("Unknown chunk type {} encountered at offset {}", type, chunkStart);
 			}
 		}
 		return pkg;
@@ -279,6 +277,13 @@ public class ResTableParser extends CommonBinaryParser implements IResParser {
 				}
 				parseEntry(pkg, id, i, config.getQualifiers());
 			}
+		}
+		if (chunkEnd > is.getPos()) {
+			// Skip remaining unknown data in this chunk (e.g. type 8 entries")
+			long skipSize = chunkEnd - is.getPos();
+			LOG.debug("Unknown data at the end of type chunk encountered, skipping {} bytes and continuing at offset {}", skipSize,
+					chunkEnd);
+			is.skip(skipSize);
 		}
 	}
 

--- a/jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java
+++ b/jadx-core/src/main/java/jadx/core/xmlgen/ResTableParser.java
@@ -150,16 +150,33 @@ public class ResTableParser extends CommonBinaryParser implements IResParser {
 		PackageChunk pkg = new PackageChunk(id, name, typeStrings, keyStrings);
 		resStorage.setAppPackage(name);
 
-		while (is.getPos() < endPos) {
+		chunkLoop: while (is.getPos() < endPos) {
 			long chunkStart = is.getPos();
 			int type = is.readInt16();
-			if (type == RES_NULL_TYPE) {
-				continue;
-			}
-			if (type == RES_TABLE_TYPE_SPEC_TYPE) {
-				parseTypeSpecChunk();
-			} else if (type == RES_TABLE_TYPE_TYPE) {
-				parseTypeChunk(chunkStart, pkg);
+			LOG.trace("res package chunk start at {} type {}", chunkStart, type);
+			switch (type) {
+				case RES_NULL_TYPE:
+					LOG.info("Null chunk type encountered at offset {}", chunkStart);
+					break;
+				case RES_TABLE_TYPE_TYPE: // 0x0201
+					parseTypeChunk(chunkStart, pkg);
+					break;
+				case RES_TABLE_TYPE_SPEC_TYPE: // 0x0202
+					parseTypeSpecChunk(chunkStart);
+					break;
+				case RES_TABLE_TYPE_LIBRARY: // 0x0203
+					parseLibraryTypeChunk(chunkStart);
+					break;
+				case RES_TABLE_TYPE_OVERLAY: // 0x0204
+					throw new IOException(
+							String.format("Encountered unsupported chunk type TYPE_OVERLAY at offset 0x%x ", chunkStart));
+				case RES_TABLE_TYPE_STAGED_ALIAS: // 0x0206
+					throw new IOException(
+							String.format("Encountered unsupported chunk type TYPE_STAGED_ALIAS at offset 0x%x ", chunkStart));
+				default:
+					// In some resources.arsc files endPos is not correctly set (e.g. points to the end of the file),
+					// so we have to stop when we encounter an unknown type entry
+					break chunkLoop;
 			}
 		}
 		return pkg;
@@ -192,16 +209,38 @@ public class ResTableParser extends CommonBinaryParser implements IResParser {
 	}
 
 	@SuppressWarnings("unused")
-	private void parseTypeSpecChunk() throws IOException {
+	private void parseTypeSpecChunk(long chunkStart) throws IOException {
 		is.checkInt16(0x0010, "Unexpected type spec header size");
-		/* int size = */
-		is.readInt32();
+		int chunkSize = is.readInt32();
+		long expectedEndPos = chunkStart + chunkSize;
 
 		int id = is.readInt8();
 		is.skip(3);
 		int entryCount = is.readInt32();
 		for (int i = 0; i < entryCount; i++) {
 			int entryFlag = is.readInt32();
+		}
+		if (is.getPos() != expectedEndPos) {
+			throw new IOException(String.format("Error reading type spec chunk at offset 0x%x", chunkStart));
+		}
+	}
+
+	private void parseLibraryTypeChunk(long chunkStart) throws IOException {
+		LOG.trace("parsing library type chunk starting at offset {}", chunkStart);
+		is.checkInt16(12, "Unexpected header size");
+		int chunkSize = is.readInt32();
+		long expectedEndPos = chunkStart + chunkSize;
+		int count = is.readInt32();
+		for (int i = 0; i < count; i++) {
+			int packageId = is.readInt32();
+			String packageName = is.readString16Fixed(128);
+			LOG.info("Found resource shared library {}, pkgId: {}", packageName, packageId);
+			if (is.getPos() > expectedEndPos) {
+				throw new IOException("reading after chunk end");
+			}
+		}
+		if (is.getPos() != expectedEndPos) {
+			throw new IOException(String.format("Error reading library chunk at offset 0x%x", chunkStart));
 		}
 	}
 

